### ISSUE TITLE
Accordion semantic variance and accessbility loss

### ISSUE THREAD

**Issue by ygatesoupe** [ng-bootstrap/ng-bootstrap#1050]
Hello, and thanks for taking the effort in porting bootstrap components to Angular 2.

My concern is on losing html semantic and aria roles in the process. As an example the accordion component :

In bootstrap 4 we have :

```html
<div id="accordion" role="tablist" aria-multiselectable="true">
  <div class="card">
    <div class="card-header" role="tab" id="headingOne">
      <h5 class="mb-0">
        <a data-toggle="collapse" data-parent="#accordion" href="#collapseOne" aria-expanded="true" aria-controls="collapseOne">
          Collapsible Group Item #1
        </a>
      </h5>
    </div>

    <div id="collapseOne" class="collapse in" role="tabpanel" aria-labelledby="headingOne">
      <div class="card-block">
        Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch. Food truck quinoa nesciunt laborum eiusmod. Brunch 3 wolf moon tempor, sunt aliqua put a bird on it squid single-origin coffee nulla assumenda shoreditch et. Nihil anim keffiyeh helvetica, craft beer labore wes anderson cred nesciunt sapiente ea proident. Ad vegan excepteur butcher vice lomo. Leggings occaecat craft beer farm-to-table, raw denim aesthetic synth nesciunt you probably haven't heard of them accusamus labore sustainable VHS.
      </div>
    </div>
  </div>
<div class="card">
    <div class="card-header" role="tab" id="headingTwo">
      <h5 class="mb-0">
        <a class="collapsed" data-toggle="collapse" data-parent="#accordion" href="#collapseTwo" aria-expanded="false" aria-controls="collapseTwo">
          Collapsible Group Item #2
        </a>
      </h5>
    </div>
    <div id="collapseTwo" class="collapse" role="tabpanel" aria-labelledby="headingTwo">
      <div class="card-block">
        Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch. Food truck quinoa nesciunt laborum eiusmod. Brunch 3 wolf moon tempor, sunt aliqua put a bird on it squid single-origin coffee nulla assumenda shoreditch et. Nihil anim keffiyeh helvetica, craft beer labore wes anderson cred nesciunt sapiente ea proident. Ad vegan excepteur butcher vice lomo. Leggings occaecat craft beer farm-to-table, raw denim aesthetic synth nesciunt you probably haven't heard of them accusamus labore sustainable VHS.
      </div>
    </div>
  </div>
</div>
```

Each panel group (title + content) is wrapped in a single card element. Aria roles are set accordingly (tablist, tab, aria-labelledby, ...)

Whereas with ng-bootstrap rendred accordion component we get something like this : 

```html
<ngb-accordion activeids="ngb-panel-0" ng-reflect-active-ids="ngb-panel-0">
  <div class="card">
      <div ng-reflect-class-name="card-header " class="card-header active">
        <a href="">Simple</a>
      </div>
      <div class="card-block">
        Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. 3 wolf moon officia
        aute, non cupidatat skateboard dolor brunch. Food truck quinoa nesciunt laborum eiusmod. Brunch 3 wolf moon tempor,
        sunt aliqua put a bird on it squid single-origin coffee nulla assumenda shoreditch et. Nihil anim keffiyeh helvetica,
        craft beer labore wes anderson cred nesciunt sapiente ea proident. Ad vegan excepteur butcher vice lomo. Leggings
        occaecat craft beer farm-to-table, raw denim aesthetic synth nesciunt you probably haven't heard of them accusamus
        labore sustainable VHS.
     </div>
      <div ng-reflect-class-name="card-header " class="card-header ">
        <a href="">
          <span>★ <b>Fancy</b> title ★</span>
        </a>
      </div>
      <div ng-reflect-class-name="card-header " class="card-header ">
        <a href="" class="text-muted">Disabled</a>
      </div>
  </div>
</ngb-accordion>
```

Aria roles are gone and all panel titles and content are wrapped in the same card element. 

I don't understand the choice of dropping accessibility since it's already here and working in bootstrap 4, and html semantic divergence can be confusing and lead to presentation bug (custom styling).

### Link to minimally-working plunker that reproduces the issue:

Bootstrap accordion example can be found [here](http://v4-alpha.getbootstrap.com/components/collapse/#accordion-example)

ng-bootstrap accordion can be found [here](http://plnkr.co/edit/PuChHuzog8YhkejLeqwa?p=preview)

### Version of Angular, ng-bootstrap, and Bootstrap:

Angular: 2.1.0

ng-bootstrap: 1.0.0-alpha.13

Bootstrap:  v4.0.0-alpha.5

**Comment by pkozlowski-opensource** [comment]
> I don't understand the choice of dropping accessibility since it's already here and working in bootstrap 4, and html semantic divergence can be confusing and lead to presentation bug (custom styling).

We are not "choosing to drop accessibility", far from it! It is just that we didn't get to complete accessibility review and fixes yet. Our thinking was / is that we want to stabilise the public APIs first and when its done add things like animations, accessibility and keyboard support (in other words all features than will make widgets solid but don't require breaking API changes).

So once again, it is not like we don't care about accessibility nor that we are not working on it deliberately. It is just that we've prioritized public APIs over other enhancements. If you want us to get there faster a PR fixing things you've noticed would be helpful.

Also please note that we didn't "remove" anythings: cards were not present in Bootstrap 3 / angular-ui/bootstrap and Bootstrap 4 added example of an accordion based on cards only very recently (after we did initial implementation of the accordion).

In short: accessibility is important for us and is part of the roadmap. Helping with PRs would be appreciated.

**Comment by ygatesoupe** [comment]
> accessibility is important for us and is part of the roadmap

Ok, I'm glad to read that. I'll fork and try to provide PR.

**Comment by pkozlowski-opensource** [comment]
@ygatesoupe I've sent PR #1070 to address accessibility-related attributes. Would be great if you could review this one.

We still got a bit of discrepancy between [Bootstrap's proposed markup](http://v4-alpha.getbootstrap.com/components/collapse/#accordion-example) and the markup we are producing:
* [ ] BS 4 is using a separate card for each panel while we have just one card.  For now I'm going to leave it as-is since I think that our look & feel is better. I will follow up with the Bootstrap team and eventually change things in this project.
* [ ] BS 4 is using `<h5>` around header links - we should probably add it as well.
* [ ] BS 4 is adding a wrapper div around `card-block` and frankly I don't see why this one is needed. I will follow up with the Bootstrap team and eventually change things in this project.

**Comment by ygatesoupe** [comment]
@pkozlowski-opensource I'm sorry I missed the notification. Thank you for you commit, it's all here and fine by me concerning aria attributes.

As for markup please consider this use case:
I have a set of bootstrap css overriding rules. If I migrate a jqurey / bootstrap project into angular2 / ng-boostrap environement keeping same set of css, accordions breaks because of different markup induced by ng-bootstrap.

In my opinion it would be better to stick to native bootstrap markup as default template, maybe providing option to set it to a different look & feel. 

`<h5>` and  content wrapper removal are less of a problem to me but could be for someone else.

**Comment by coutin** [comment]
Hello, 

Just saw the pull request and adding aria is great.
I see that one aria attribute is missing it's about the aria-selected is missing; The WAI-ARIA specification https://www.w3.org/TR/wai-aria/states_and_properties#aria-selected states that using aria-mutliselectable attribute should provide an aria-selected attibute to every selectable containers.

Will probably propose a PR for that.

**Comment by thbt** [comment]
@pkozlowski-opensource apparently the use of the card-component is necessary to extend the collapse behavior to the accordion, according to this [issue](https://github.com/twbs/bootstrap/pull/21746)

I'll put up a PR tomorrow morning.

**Comment by thbt** [comment]
I was thinking it could be done differently : 
If the panel title template is not set, we display the title using the correct bootstrap markdown.
If the panel title template is set, we display it.
Something like this : 
```
<h5 class="mb-0" *ngIf="!panel.titleTpl">
  <a href (click)="!!toggle(panel.id)" (focus)="panel.focused = true"
    (blur)="panel.focused = false" [class.text-muted]="panel.disabled"
    [attr.aria-expanded]="isOpen(panel.id)" [attr.aria-controls]="panel.id">
    {{panel.title}}
  </a>
</h5>
<template *ngIf="!panel.titleTpl" [ngTemplateOutlet]="panel.titleTpl?.templateRef"></template>
``` 
Obviously `NgbPanelTitle` will be modified to add the mandatory aria attributes to it's host.

What do you think ?

**Comment by pkozlowski-opensource** [comment]
Please note that Bootstrap 4 is changing markup for accordions (not using cards any more), see:
* https://github.com/twbs/bootstrap/pull/22138
* https://github.com/twbs/bootstrap/pull/22251

We should wait till the situation with accordion's markup stabilizes before doing any more work here. 

Marking as blocked on https://github.com/twbs/bootstrap/pull/22251
